core/src/main/scala/org/http4s/Query.scala

```diff
@@ -1,13 +1,82 @@
 package org.http4s
 
-import org.http4s.Query._
+import org.http4s.FormQuery._
 import org.http4s.parser.QueryParser
-import org.http4s.util.{UrlFormCodec, UrlCodingUtils, Writer, Renderable}
+import org.http4s.util._
+import org.http4s.util.encoding.UriCodingUtils
 
 import scala.collection.generic.CanBuildFrom
-import scala.collection.immutable.{BitSet, IndexedSeq}
+import scala.collection.immutable.IndexedSeq
 import scala.collection.mutable.ListBuffer
-import scala.collection.{ IndexedSeqOptimized, mutable }
+import scala.collection.{IndexedSeqOptimized, mutable}
+
+
+sealed trait Query extends QueryOps with Renderable {
+  def asForm: FormQuery
+  def isEmpty: Boolean
+  def encoded: String
+}
+
+/** Represents the absence of a query component */
+object EmptyQuery extends Query {
+  override def asForm: FormQuery = FormQuery()
+  override def encoded: String = "<please check isEmpty before using this value>"
```

Comment:
We have `org.http4s.util.bug` for cases like this.  But maybe introducing this method on `NonEmptyQuery` is better?